File: CS/Scaffolding.CustomFilter/ViewModels/Issue/IssueCollectionViewModel.Extensions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Linq.Expressions;
using Scaffolding.CustomFilter.Model;

namespace Scaffolding.CustomFilter.ViewModels {
    public partial class IssueCollectionViewModel {
        static readonly IEnumerable<FilterExpressionInfo<Issue>> filters = new FilterExpressionInfo<Issue>[] {
                        new FilterExpressionInfo<Issue> ("All",  x => true ),
                        new FilterExpressionInfo<Issue> ("For support", x => x.IssueType == IssueType.Question ),
                        new FilterExpressionInfo<Issue> ("For developers", x => x.IssueType == IssueType.Bug || x.IssueType == IssueType.Suggestion ),
                    };
        public IEnumerable<FilterExpressionInfo<Issue>> Filters { get { return filters; } }
        protected override void OnInitializeInRuntime() {
            base.OnInitializeInRuntime();
            FilterExpression = Filters.First().Expression;
        }
    }
    public class FilterExpressionInfo<TEntity> {
        public FilterExpressionInfo(string displayName, Expression<Func<TEntity, bool>> expression) {
            Expression = expression;
            DisplayName = displayName;
        }
        public Expression<Func<TEntity, bool>> Expression { get; private set; }
        public string DisplayName { get; private set; }
    }
}

File: VB/Scaffolding.CustomFilter/ViewModels/Issue/IssueCollectionViewModel.Extensions.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Linq.Expressions
Imports Scaffolding.CustomFilter.Model
Imports DevExpress.Mvvm

Namespace Scaffolding.CustomFilter.ViewModels
	Partial Public Class IssueCollectionViewModel
		Implements ISupportLogicalLayout(Of String)

'INSTANT VB NOTE: The variable filters was renamed since Visual Basic does not allow variables and other class members to have the same name:
		Private Shared ReadOnly filters_Renamed As IEnumerable(Of FilterExpressionInfo(Of Issue)) = New FilterExpressionInfo(Of Issue)() {
			New FilterExpressionInfo(Of Issue) ("All", Function(x) True),
			New FilterExpressionInfo(Of Issue) ("For support", Function(x) x.IssueType = IssueType.Question),
			New FilterExpressionInfo(Of Issue) ("For developers", Function(x) x.IssueType = IssueType.Bug OrElse x.IssueType = IssueType.Suggestion)
		}
		Public ReadOnly Property Filters() As IEnumerable(Of FilterExpressionInfo(Of Issue))
			Get
				Return filters_Renamed
			End Get
		End Property

		Protected Overrides Sub OnInitializeInRuntime()
			MyBase.OnInitializeInRuntime()
			FilterExpression = Filters.First().Expression
		End Sub

		Private Sub ISupportLogicalLayoutGeneric_RestoreState(ByVal state As String) Implements ISupportLogicalLayout(Of String).RestoreState
			FilterExpression = Filters.First(Function(f) f.DisplayName = state).Expression
		End Sub

		Private Function ISupportLogicalLayoutGeneric_SaveState() As String Implements ISupportLogicalLayout(Of String).SaveState
			Return Filters.First(Function(f) Equals(f.Expression, FilterExpression)).DisplayName
		End Function
	End Class
	Public Class FilterExpressionInfo(Of TEntity)
		Public Sub New(ByVal displayName As String, ByVal expression As Expression(Of Func(Of TEntity, Boolean)))
			Me.Expression = expression
			Me.DisplayName = displayName
		End Sub
		Private privateExpression As Expression(Of Func(Of TEntity, Boolean))
		Public Property Expression() As Expression(Of Func(Of TEntity, Boolean))
			Get
				Return privateExpression
			End Get
			Private Set(ByVal value As Expression(Of Func(Of TEntity, Boolean)))
				privateExpression = value
			End Set
		End Property
		Private privateDisplayName As String
		Public Property DisplayName() As String
			Get
				Return privateDisplayName
			End Get
			Private Set(ByVal value As String)
				privateDisplayName = value
			End Set
		End Property
	End Class
End Namespace